archunit/src/main/java/com/tngtech/archunit/lang/conditions/SubEvents.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2014-2022 TNG Technology Consulting GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.tngtech.archunit.lang.conditions;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+import com.tngtech.archunit.lang.ConditionEvent;
+import com.tngtech.archunit.lang.ConditionEvents;
+
+final class SubEvents implements ConditionEvents {
```

Comment:
Some thoughts about naming:

Do you think that `SubEvents` is a good name? (I didn't understand it.)
How about `InvertableConditionEvents` (as this class tracks `allowedEvents`, too)?

Similarly, I'd probably rename `ConditionEventsInternal` to something like `SimpleConditionEvents` or `LeanConditionEvents` and leave a comment that this class only tracks violations.

I was at first surprised that `ConditionEvents.Factory.create()` builds lean `ConditionEvents`, but then realized that this was okay as the `interface ConditionEvents` has changed and now really only deals with violations. So `ConditionEvents` might not be the best name any more: IMO it's rather a `ViolationsContainer` now, but I guess that you didn't want to break _all_ existing `ArchCondition`s (which have this type in the signature of the `check` method)... 😉

Reply:
But it's not completely true that it is always only a `ViolationsContainer` as long as you can add `ConditionEvents` with `satisfied == true`, no? I mean, you can always add satisfied and violated `ConditionEvents`, so I still think `ConditionEvents` is a fair name. What the interface now does not allow anymore is to *retrieve* informations about satisfied events. But it can still transport them. Nevertheless, since we can't *retrieve* satisfied `ConditionEvents` or are influenced in any way on the public surface, we don't need to track allowed events anymore in the standard case :man_shrugging: But I could imagine users doing the same as ArchUnit does internally, i.e. use their own implementation to track or handle more :man_shrugging:

Reply:
About the other naming, like `SubEvents`, I have to admit that I haven't spent too much brain power on the naming as it was never public API :see_no_evil: For me the name came from "I'm collecting a subset of events with this container, then transform it and add it to the parent container". But earlier it was local variables that I called `subEvents`, not a whole type...

Reply:
About `Lean` I'm not a super big fan, because that for me also doesn't convey much information :man_shrugging: Or in other words, it's quite fuzzy to me what this means...

Reply:
The suffix `Internal` came from the convention we have for the syntax. I.e. public API is an interface, like `ConditionEvents.Factory.create()` in this case, and the implementation was usually called `xxxInternal` there. But yeah, maybe `SimpleConditionEvents` is fair, since it's the most straight forward implementation within ArchUnit without any extra properties / responsibilities.

Reply:
Right, the naming does not surface the public API.
I'd still find it useful to have a clear semantics for ArchUnit's internal types, but a code comment or JavaDoc would be enough.

Reply:
The difference to the syntax API with the `Internal` classes is that there is only one implementation.
Here we have two, so an ArchUnit developer needs to understand their semantics.

Reply:
Unfortunately, `InvertibleConditionEvents` also sounds weird to me, because every `ConditionEvent` has an `invert()` method and is thus "invertible" :cry: But it's really hard to come up with a good name. Maybe in the end it will just be `Extended...` or something fuzzy like that :disappointed: But at least it's no API, so the name is irrelevant for users...

Reply:
As discussed I'll rename `SubEvents` to `ViolatedAndSatisfiedConditionEvents` with an additional Javadoc. It's not the prettiest name, but at least it tells exactly what's the point :wink: And it's luckily no public API in the end :man_shrugging: